File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.py
def solution(numbers, hand):
    answer = ''
    keypad = dict()
    keypad[1] = (0, 3)
    keypad[2] = (1, 3)
    keypad[3] = (2, 3)
    keypad[4] = (0, 2)
    keypad[5] = (1, 2)
    keypad[6] = (2, 2)
    keypad[7] = (0, 1)
    keypad[8] = (1, 1)
    keypad[9] = (2, 1)
    keypad[0] = (1, 0)
    keypad['*'] = (0, 0)
    keypad['#'] = (2, 0)
    
    left_cur = (0, 0)
    right_cur = (2, 0)
    
    for num in numbers:
        if num in [1, 4, 7]:
            answer += 'L'
            left_cur = keypad[num]
        elif num in [3, 6, 9]:
            answer += 'R'
            right_cur = keypad[num]
        else:
            left_diff = abs(keypad[num][0] - left_cur[0]) + abs(keypad[num][1] - left_cur[1])
            right_diff = abs(keypad[num][0] - right_cur[0]) + abs(keypad[num][1] - right_cur[1])
            
            if left_diff < right_diff:
                answer += 'L'
                left_cur = keypad[num]
            elif right_diff < left_diff:
                answer += 'R'
                right_cur = keypad[num]
            else:
                if hand == 'right':
                    answer += 'R'
                    right_cur = keypad[num]
                else: 
                    answer += 'L'
                    left_cur = keypad[num]
                    
    return answer

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.java
class Solution {
    public String solution(int[] numbers, String hand) {
        StringBuilder sb = new StringBuilder();
        int leftCur = 10, rightCur = 12;
        
        for (int n : numbers) {
            
            switch (n) {
                case 1:
                case 4:
                case 7:
                    sb.append("L");
                    leftCur = n;
                    break;
                case 3:
                case 6:
                case 9:
                    sb.append("R");
                    rightCur = n;
                    break;
                default:
                    if (n == 0) n = 11;
                    
                    int x = (n-1) / 3, y = (n-1) % 3;
                    int lx = (leftCur-1) / 3, ly = (leftCur-1) % 3;
                    int rx = (rightCur-1) / 3, ry = (rightCur-1) % 3;
                    int distL = Math.abs(lx-x) + Math.abs(ly - y);
                    int distR = Math.abs(rx-x) + Math.abs(ry - y);
                    
                    char tmp = 'a';
                    if (distL < distR) {
                        tmp = 'L';
                        leftCur = n;
                    } else if (distR < distL) {
                        tmp = 'R';
                        rightCur = n;
                    } else {
                        if (hand.equals("left")) {
                            tmp = 'L';
                            leftCur = n;
                        } else {
                            tmp = 'R';
                            rightCur = n;
                        }
                    }
                    sb.append(tmp);
                    break;
            }
        }
        return sb.toString();
    }
}